jxl/src/frame/modular/predict.rs

```diff
@@ -152,17 +155,244 @@ impl Predictor {
     }
 }
 
+const NUM_PREDICTORS: usize = 4;
+const PRED_EXTRA_BITS: i64 = 3;
+const PREDICTION_ROUND: i64 = ((1 << PRED_EXTRA_BITS) >> 1) - 1;
+// Allows to approximate division by a number from 1 to 64.
+//  for (int i = 0; i < 64; i++) divlookup[i] = (1 << 24) / (i + 1);
+const DIVLOOKUP: [u32; 64] = [
+    16777216, 8388608, 5592405, 4194304, 3355443, 2796202, 2396745, 2097152, 1864135, 1677721,
+    1525201, 1398101, 1290555, 1198372, 1118481, 1048576, 986895, 932067, 883011, 838860, 798915,
+    762600, 729444, 699050, 671088, 645277, 621378, 599186, 578524, 559240, 541200, 524288, 508400,
+    493447, 479349, 466033, 453438, 441505, 430185, 419430, 409200, 399457, 390167, 381300, 372827,
+    364722, 356962, 349525, 342392, 335544, 328965, 322638, 316551, 310689, 305040, 299593, 294337,
+    289262, 284359, 279620, 275036, 270600, 266305, 262144,
+];
+
+fn add_bits<T: num_traits::PrimInt>(x: T) -> i64 {
+    x.to_i64().unwrap() << PRED_EXTRA_BITS
+}
+
+fn error_weight(x: u32, maxweight: u32) -> u32 {
+    let shift = floor_log2_nonzero(x + 1) as i32 - 5;
+    if shift < 0 {
+        4u32.wrapping_add(maxweight.wrapping_mul(DIVLOOKUP[x as usize]))
+    } else {
+        4u32.wrapping_add((maxweight.wrapping_mul(DIVLOOKUP[x as usize >> shift])) >> shift)
+    }
+}
+
+fn weighted_average(pixels: &[i64; NUM_PREDICTORS], weights: &mut [u32; NUM_PREDICTORS]) -> i64 {
+    let log_weight = floor_log2_nonzero(weights.iter().fold(0u32, |sum, el| sum.wrapping_add(*el)));
+    let weight_sum = weights.iter_mut().fold(0, |sum, el| {
+        *el >>= log_weight - 4;
+        sum + *el
+    });
+    let sum = weights
+        .iter()
+        .enumerate()
+        .fold(((weight_sum >> 1) - 1) as i64, |sum, (i, weight)| {
+            sum + pixels[i] * *weight as i64
+        });
+    (sum.wrapping_mul(DIVLOOKUP[(weight_sum - 1) as usize] as i64)) >> 24
+}
+
 #[derive(Debug)]
-pub struct WeightedPredictorState;
+pub struct WeightedPredictorState<'a> {
+    prediction: [i64; NUM_PREDICTORS],
+    pred: i64,
+    pred_errors: [Vec<u32>; NUM_PREDICTORS],
+    error: Vec<i64>,
+    header: &'a GroupHeader,
+}
+
+impl<'a> WeightedPredictorState<'a> {
+    pub fn new(header: &'a GroupHeader, xsize: usize) -> WeightedPredictorState<'a> {
+        let num_errors = (xsize + 2) * 2;
+        WeightedPredictorState {
+            prediction: [0; NUM_PREDICTORS],
+            pred: 0,
+            pred_errors: from_fn(|_| vec![0; num_errors]),
+            error: vec![0; num_errors],
+            header,
+        }
+    }
 
-impl WeightedPredictorState {
-    pub fn new(_header: &GroupHeader) -> Self {
-        // TODO(veluca): implement the weighted predictor.
-        Self
+    #[allow(dead_code)]
+    pub fn update_errors(&mut self, correct_val: i64, pos: (usize, usize), xsize: usize) {
```

Comment:
Can you add the call to `update_errors` too?

Reply:
I'm not sure where it should be called - could you point it out?

Reply:
Done